client/internal/httpsender_test.go

```diff
@@ -98,3 +100,50 @@ EKTcWGekdmdDPsHloRNtsiCa697B2O9IFA==
 
 	return cert, nil
 }
+
+func TestHTTPSenderRetryForFailedRequests(t *testing.T) {
+
+	srv := StartMockServer(t)
+	srv.srv.Config.ReadTimeout = 1 * time.Nanosecond
```

Comment:
I wonder how reliable this is. There is still a small chance that this may succeed.

Instead can we simulate an error by attempting to connect to a non-existing server? So make the first request attempt before starting the mock server. That should fail with an error. Then start them mock server.

Reply:
>I wonder how reliable this is. There is still a small chance that this may succeed.

It was pretty reliable  (observation from running it on a local machine a few dozen times).

> Instead can we simulate an error by attempting to connect to a non-existing server? So make the first request attempt before starting the mock server. That should fail with an error. Then start them mock server

I thought about this. Once the request gets prepared and goes into the retry loop, I couldn't think of a clean way to change the url. Any ideas?

Reply:
Is it necessary to change the URL? You can find a free port number nnnn, hit http://localhost:nnnn with a request and provided that the mock server is not started the request should fail with an error. Then start the server and try the same URL again, it should pass.

Reply:
Let me check that.

Reply:
Connection to a non-existing server wasn't enough. I updated the test to take control of the underlying conn and simulate the error. Here is the run with a failed test without the fix: https://github.com/open-telemetry/opamp-go/actions/runs/6085619221/job/16510155045 and run that passes with fix https://github.com/open-telemetry/opamp-go/actions/runs/6085624937/job/16510179749. Let me know what do you think.